File: swift_reference/Models/WeatherData.swift
import Foundation

struct WeatherResponse: Codable {
    let latitude: Double
    let longitude: Double
    let current: CurrentWeather?
    let daily: DailyWeather?

    enum CodingKeys: String, CodingKey {
        case latitude, longitude, current, daily
    }
}

struct CurrentWeather: Codable {
    let time: String
    let temperature2m: Double
    let relativeHumidity2m: Int
    let apparentTemperature: Double
    let precipitation: Double
    let weatherCode: Int
    let windSpeed10m: Double
    let windDirection10m: Int
    let windGusts10m: Double

    enum CodingKeys: String, CodingKey {
        case time
        case temperature2m = "temperature_2m"
        case relativeHumidity2m = "relative_humidity_2m"
        case apparentTemperature = "apparent_temperature"
        case precipitation
        case weatherCode = "weather_code"
        case windSpeed10m = "wind_speed_10m"
        case windDirection10m = "wind_direction_10m"
        case windGusts10m = "wind_gusts_10m"
    }
}

struct DailyWeather: Codable {
    let time: [String]
    let weatherCode: [Int]
    let temperature2mMax: [Double]
    let temperature2mMin: [Double]
    let precipitationSum: [Double]
    let precipitationProbabilityMax: [Int]
    let windSpeed10mMax: [Double]
    let windGusts10mMax: [Double]

    enum CodingKeys: String, CodingKey {
        case time
        case weatherCode = "weather_code"
        case temperature2mMax = "temperature_2m_max"
        case temperature2mMin = "temperature_2m_min"
        case precipitationSum = "precipitation_sum"
        case precipitationProbabilityMax = "precipitation_probability_max"
        case windSpeed10mMax = "wind_speed_10m_max"
        case windGusts10mMax = "wind_gusts_10m_max"
    }
}

struct DayForecast: Identifiable {
    let id = UUID()
    let date: Date
    let weatherCode: Int
    let highTemp: Double
    let lowTemp: Double
    let precipitationSum: Double
    let precipitationProbability: Int
    let maxWindSpeed: Double
    let maxWindGusts: Double

    var weatherDescription: String {
        WeatherCode.description(for: weatherCode)
    }

    var weatherIcon: String {
        WeatherCode.sfSymbol(for: weatherCode)
    }

    var isSevere: Bool {
        weatherCode >= 95 || maxWindGusts > 50 || precipitationSum > 25
    }
}

enum WeatherCode {
    static func description(for code: Int) -> String {
        switch code {
        case 0: return "Clear sky"
        case 1: return "Mainly clear"
        case 2: return "Partly cloudy"
        case 3: return "Overcast"
        case 45, 48: return "Foggy"
        case 51, 53, 55: return "Drizzle"
        case 56, 57: return "Freezing drizzle"
        case 61, 63, 65: return "Rain"
        case 66, 67: return "Freezing rain"
        case 71, 73, 75: return "Snow"
        case 77: return "Snow grains"
        case 80, 81, 82: return "Rain showers"
        case 85, 86: return "Snow showers"
        case 95: return "Thunderstorm"
        case 96, 99: return "Thunderstorm with hail"
        default: return "Unknown"
        }
    }

    static func sfSymbol(for code: Int) -> String {
        switch code {
        case 0: return "sun.max.fill"
        case 1: return "sun.min.fill"
        case 2: return "cloud.sun.fill"
        case 3: return "cloud.fill"
        case 45, 48: return "cloud.fog.fill"
        case 51, 53, 55: return "cloud.drizzle.fill"
        case 56, 57: return "cloud.sleet.fill"
        case 61, 63, 65: return "cloud.rain.fill"
        case 66, 67: return "cloud.sleet.fill"
        case 71, 73, 75: return "cloud.snow.fill"
        case 77: return "cloud.snow.fill"
        case 80, 81, 82: return "cloud.heavyrain.fill"
        case 85, 86: return "cloud.snow.fill"
        case 95: return "cloud.bolt.fill"
        case 96, 99: return "cloud.bolt.rain.fill"
        default: return "questionmark"
        }
    }
}

File: lib/models/weather_data.dart
class WeatherResponse {
  final double latitude;
  final double longitude;
  final CurrentWeather? current;
  final DailyWeather? daily;

  WeatherResponse({
    required this.latitude,
    required this.longitude,
    this.current,
    this.daily,
  });

  factory WeatherResponse.fromJson(Map<String, dynamic> json) {
    return WeatherResponse(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      current: json['current'] != null
          ? CurrentWeather.fromJson(json['current'])
          : null,
      daily: json['daily'] != null ? DailyWeather.fromJson(json['daily']) : null,
    );
  }
}

class CurrentWeather {
  final String time;
  final double temperature2m;
  final int relativeHumidity2m;
  final double apparentTemperature;
  final double precipitation;
  final int weatherCode;
  final double windSpeed10m;
  final int windDirection10m;
  final double windGusts10m;

  CurrentWeather({
    required this.time,
    required this.temperature2m,
    required this.relativeHumidity2m,
    required this.apparentTemperature,
    required this.precipitation,
    required this.weatherCode,
    required this.windSpeed10m,
    required this.windDirection10m,
    required this.windGusts10m,
  });

  factory CurrentWeather.fromJson(Map<String, dynamic> json) {
    return CurrentWeather(
      time: json['time'] as String,
      temperature2m: (json['temperature_2m'] as num).toDouble(),
      relativeHumidity2m: json['relative_humidity_2m'] as int,
      apparentTemperature: (json['apparent_temperature'] as num).toDouble(),
      precipitation: (json['precipitation'] as num).toDouble(),
      weatherCode: json['weather_code'] as int,
      windSpeed10m: (json['wind_speed_10m'] as num).toDouble(),
      windDirection10m: json['wind_direction_10m'] as int,
      windGusts10m: (json['wind_gusts_10m'] as num).toDouble(),
    );
  }
}

class DailyWeather {
  final List<String> time;
  final List<int> weatherCode;
  final List<double> temperature2mMax;
  final List<double> temperature2mMin;
  final List<double> precipitationSum;
  final List<int> precipitationProbabilityMax;
  final List<double> windSpeed10mMax;
  final List<double> windGusts10mMax;

  DailyWeather({
    required this.time,
    required this.weatherCode,
    required this.temperature2mMax,
    required this.temperature2mMin,
    required this.precipitationSum,
    required this.precipitationProbabilityMax,
    required this.windSpeed10mMax,
    required this.windGusts10mMax,
  });

  factory DailyWeather.fromJson(Map<String, dynamic> json) {
    return DailyWeather(
      time: List<String>.from(json['time']),
      weatherCode: List<int>.from(json['weather_code']),
      temperature2mMax: List<double>.from(
          (json['temperature_2m_max'] as List).map((e) => (e as num).toDouble())),
      temperature2mMin: List<double>.from(
          (json['temperature_2m_min'] as List).map((e) => (e as num).toDouble())),
      precipitationSum: List<double>.from(
          (json['precipitation_sum'] as List).map((e) => (e as num).toDouble())),
      precipitationProbabilityMax:
          List<int>.from(json['precipitation_probability_max']),
      windSpeed10mMax: List<double>.from(
          (json['wind_speed_10m_max'] as List).map((e) => (e as num).toDouble())),
      windGusts10mMax: List<double>.from(
          (json['wind_gusts_10m_max'] as List).map((e) => (e as num).toDouble())),
    );
  }
}

class DayForecast {
  final String id;
  final DateTime date;
  final int weatherCode;
  final double highTemp;
  final double lowTemp;
  final double precipitationSum;
  final int precipitationProbability;
  final double maxWindSpeed;
  final double maxWindGusts;

  DayForecast({
    required this.id,
    required this.date,
    required this.weatherCode,
    required this.highTemp,
    required this.lowTemp,
    required this.precipitationSum,
    required this.precipitationProbability,
    required this.maxWindSpeed,
    required this.maxWindGusts,
  });

  bool get isSevere =>
      weatherCode >= 95 || maxWindGusts > 50 || precipitationSum > 25;
}
